Expand documentation for `DeliverySubject` (#426)
nats-io/nats.docs: nats-concepts/jetstream/consumers.md

## DeliverySubject

The delivery subject is used by NATS to orchestrate how messages reach subscriptions. When a message is published to a JetStream, NATS will know of all consumers that have registered an interested in that message. It will then deliver the message to each consumer via their `DeliverySubject`. In this sense, a `DeliverySubject` can be considered like an "inbox" for each subscription.

> For this reason, `DeliverySubject` is not allowed for pull subscriptions, since in that case the subscription is fetching the messages for itself. Similarly, a delivery subject is required for all push subscriptions, as well as for queue subscribing as it configures a subject that all the queue consumers should listen on.

Reviewer: ```suggestion
`DeliverSubject` is not applicable for pull consumers, since a pull-bound subscription is responsible for *fetching* the messages for itself.
```

> The delivery subject should not be confused with the stream's subject. In fact, if a delivery subject had the same value as a stream's subject, it could create an infinite loop as NATS would re-deliver messages to the same subjects again and again. NATS will prevent the creation of this by emitting the following error: `Consumer creation failed: consumer deliver subject forms a cycle (10081)`.

Reviewer: ```suggestion
The deliver subject should not be confused with the stream's subject. In fact, if a deliver subject had the same value as a stream's subject, NATS will emit the following error: `Consumer creation failed: consumer deliver subject forms a cycle (10081)` to prevent published messages from being delivered on an overlapping subject.
```

> If two consumers share the same `DeliverySubject` then the messages will be delivered to both, since subscriptions will be bound to the same subject. In cases in which you want a subscription to receive all (filtered) messages of a stream, a unique `DeliverySubject` should be provided.

Reviewer 2: Perhaps it should be added that when two consumers share the same `DeliverySubject`, and in case where one consumer responds with `NAK` the message will not be re-delivered if the other consumer responded with `ACK`.

> If two consumers share the same `DeliverySubject` then the messages will be delivered to both, since subscriptions will be bound to the same subject. In cases in which you want a subscription to receive all (filtered) messages of a stream, a unique `DeliverySubject` should be provided.

Reviewer: ```suggestion
If two consumers share the same `DeliverSubject` then the messages will be delivered twice, since subscriptions will be bound to the same subject. In cases in which you want a subscription to receive all (filtered) messages of a stream, a unique `DeliverSubject` should be provided.
```
